Leak Caffe2 threadpool in child processes right after fork to prevent segfault

imaginary-person opened the pull request:
## Problem summary
Fixes #54752 - when the number of threads is more than 3 and at least one `set_num_threads` invocation has taken place before forking child processes by the dataloader, `set_num_threads(1)` in the child process causes a segfault, as during its invocation, the child process is made to handle the data structures of the Caffe2 thread-pool of the parent process, whose data structures it inherits from the parent process (these threads don't exist in the child process, but some of its data structures do, due to the copy-on-write technique used by `fork`).


## Solution
@malfet [advised](https://github.com/pytorch/pytorch/issues/54752#issuecomment-810315302) & [authored code](https://github.com/pytorch/pytorch/pull/54895#pullrequestreview-625670122) for adding a `pthread_atfork` handler in `pytorch/caffe2/utils/threadpool/pthreadpool-cpp.cc`, that's invoked in the child process right after fork, to leak the Caffe2 thread-pool (the child inherits the thread-pool's data structures from its parent process, but doesn't actually have those threads, since after `fork` , a child process only has one thread).


## Additional changes
Added unittest `test_no_segfault` to test for this issue in `test_dataloader.py` 
Also enabled `test_segfault` (which actually makes sure that segfaults happen in worker processes in a particular case).

imaginary-person commented:
Hello @ezyang @malfet,
Do you think such a solution would be acceptable towards fixing #54752?
While the CI tests are still running, I only tested `test_dataloader.py` & `test_utils.py` locally.

Thank you!

imaginary-person commented:
I should probably add a test.

ezyang commented:
Does the test fail prior to this change and pass after it?

While this looks plausible, I do agree that it seems kind of bad to thrash the thread pool every time you construct the dataloader. Maybe @malfet has more comments.

imaginary-person commented:
Thank you for taking a look, @ezyang!

> Does the test fail prior to this change and pass after it?

I had later discovered that this test doesn't fail without this change, but I'm waiting for the tests to complete, before adding a test  based on the example code in the issue. I'm sorry, I should've edited the comment.

imaginary-person commented on caffe2/utils/threadpool/pthreadpool-cpp.cc:
```diff
@@ -2,8 +2,42 @@
+
+namespace {
+// Number of threads in the thread-pool before fork
+std::atomic<int> num_threads_before_fork{-1};
```
I literally copy-pasted this from `ParallelOpenMP.cpp`. I'm not sure but I think a simple `size_t` would have sufficed.

imaginary-person commented on caffe2/utils/threadpool/pthreadpool-cpp.cc:
```diff
@@ -2,8 +2,42 @@
+// segfault, as they try to handle inherited data-structures of parent's Caffe2
+// thread-pool, although those threads don't exist in child processes.
+// Ref: https://github.com/pytorch/pytorch/issues/54752#issuecomment-810315302
+void before_fork() {
```
These handlers should probably be `static`. I'll modify after CI checks would complete.

imaginary-person commented on caffe2/utils/threadpool/pthreadpool-cpp.cc:
```diff
@@ -59,6 +93,12 @@ size_t getDefaultNumThreads();
+    // Register pthread_atfork handlers to prevent segfaults in worker processes
+    // after fork.
+    static std::once_flag flag;
+    std::call_once(flag, []{ pthread_atfork(before_fork, after_fork, NULL); });
```
At 3 places in the codebase, `pthread_atfork` is being used (for the child process handler, which won't conflict with this one), and 2 of them use `pthread_atfork` in `std::call_once`, but I'm not sure if it's needed here as well.

imaginary-person committed:
pthread_atfork and unittest

imaginary-person commented:
If CI tests would fail, I would add add `std::call_once` to the `pthread_atfork` invocation.

malfet commented on caffe2/utils/threadpool/pthreadpool-cpp.cc:
```diff
@@ -2,8 +2,42 @@
+// Ref: https://github.com/pytorch/pytorch/issues/54752#issuecomment-810315302
+static void before_fork() {
+#ifdef USE_PTHREADPOOL
+  PThreadPool* const pool = caffe2::pthreadpool();
```
This call would create `pthreadpool` even if it was used before fork, wouldn't it?

imaginary-person replied:
Yes. Thanks a lot for your help! I'll make the necessary changes.

malfet reviewed (changes requested):
Change in its current form, would cause an unnecessary thread pool creation before/after fork. A "better" (i.e. less intrusive solution would be to leak the existing pool after the fork using  https://en.cppreference.com/w/cpp/memory/unique_ptr/release, i.e. do something like:
```
static std::atomic<bool> leak_corrupted_threadpool = false;
void child_atfork() {
   leak_corrupted_threadpool = true;
}
PThreadPool* pthreadpool() {
  static auto threadpool = std::make_unique<PThreadPool>(getDefaultNumThreads());
  std::once(oncep, []() {
    pthread_atfork(nullptr, nullptr, child_atfork);
  });
 if (leak_corrupted_threadpool.compare_exchange_weak(true, false)) {
   if (auto leaked = pthreadpool.release()) {
     auto num_threads = leaked->get_thread_count();
    threadpool.reset(new PThreadPool(num_threads));
    TORCH_WARN("Leaking thread pool before rok
   }
 }
 return threadpool.get();
}
```

```

malfet commented on caffe2/utils/threadpool/pthreadpool-cpp.cc:
```diff
@@ -59,6 +93,11 @@ size_t getDefaultNumThreads();
+#ifndef WIN32
+  // Register pthread_atfork handlers to prevent segfaults in worker processes
+  // after fork.
+  pthread_atfork(before_fork, after_fork, nullptr);
```
This would register new callback every time this function is called, please used std::once

imaginary-person committed:
Add code provided by malfet

Replaced all code with the code provided by malfet, except for replacing compare_exchange_weak with as compare_exchange_strong, as the latter might cause an issue on NUMA architectures

imaginary-person committed:
Revise comment

imaginary-person committed:
Fix bug

imaginary-person committed:
remove static keyword for true_bool

imaginary-person commented:
The CI failure is unrelated.

imaginary-person committed:
[skip ci] revise warning

[skip ci] Only revised the warning

malfet reviewed (approved):
Thank you very much for working on the fix across numerous iterations of the PR.

malfet commented on caffe2/utils/threadpool/pthreadpool-cpp.cc:
```diff
@@ -57,8 +72,22 @@ void PThreadPool::run(
+  });
+#endif
+  auto true_bool = true;
+  if (leak_corrupted_threadpool.compare_exchange_strong(true_bool, false)) {
```
@robieta is there an easy way to measure how much this check affects perf and how frequently this function is called?

I also wonder if we are likely to race here, and if global bool would be sufficient here and perhaps just annotated with `C10_UNLIKELY`

But all that can probably be done in followup PR

imaginary-person replied:
@malfet, thanks for your guidance with this PR!

`std::atomic<bool>` is probably an overkill, so I can change it here if you'd like. 
I had used it because I don't understand the thread-model of PyTorch well, and since `num_threads` in `ParallelOpenMP.cpp` is atomic, I wasn't sure if there could be a race condition here as well. 
It'd be great if you could clarify why we use an atomic variable for `num_threads` in `ParallelOpenMP.cpp`, because I couldn't figure out what race-condition could arise there. I searched old issues & PRs, but couldn't find a reference. Thanks!

ezyang replied:
At the very least, I think a weak CEX may be good enough. I bet @swolchok would know off the top of his head though.

imaginary-person replied:
Hello @ezyang, do you think it's possible for weak CEX to cause an issue on NUMA nodes? Thanks!

swolchok replied:
https://en.cppreference.com/w/cpp/atomic/atomic/compare_exchange says that when weak has to go in a loop and strong wouldn't, strong is better. (Weak is allowed to fail spuriously.)

imaginary-person replied:
Hello @swolchok, do you think there could be a race-condition here if `leak_corrupted_threadpool` wasn't atomic, but simply a bool, and `C10_UNLIKELY` was used for it instead?

I used an atomic type for `leak_corrupted_threadpool` here, just to be on the safe side, because I don't understand the PyTorch thread-model well. Thank you!

swolchok replied:
I think this is already racy as written with the CAS, let alone if you use the corrupted thread bool. Ordering of events:

Thread 1 enters the CAS
Thread 2 enters the CAS
Thread 1 sees true and writes false
Thread 2 sees false
thread 1 calls threadpool.release()
Thread 2 executes threadpool.get()

bam, unsynchronized access to threadpool where one thread writes and one thread reads. Offhand I think you need either a lock or atomics on the threadpool pointer itself.

I don't see any restrictions on which threads may call `pthreadpool()` and it's called via general-looking primitives like parallel_for in ParallelNative.h.

swolchok replied:
I think this is already racy as written with the CAS, let alone if you use a non-atomic bool. Ordering of events:

Thread 1 enters the CAS
Thread 2 enters the CAS
Thread 1 sees true and writes false
Thread 2 sees false
thread 1 calls threadpool.release()
Thread 2 executes threadpool.get()

bam, unsynchronized access to threadpool where one thread writes and one thread reads. Offhand I think you need either a lock (which works via CAS internally anyway) or atomics on the threadpool pointer itself.

I don't see any restrictions on which threads may call pthreadpool() and it's called via general-looking primitives like parallel_for in ParallelNative.h.

imaginary-person replied:
Thanks for the clarification, @swolchok!

Now that I think of it, this race-condition won't arise because `leak_corrupted_threadpool` would be set to true only after a fork, when the process would have one thread (after fork). All the threads would always see `leak_corrupted_threadpool` as false.

imaginary-person replied:
> I also wonder if we are likely to race here, and if global bool would be sufficient here and perhaps just annotated with `C10_UNLIKELY`
But all that can probably be done in followup PR

@malfet, I think you were right. Opened #55341 to make `leak_corrupted_threadpool` a global bool.

imaginary-person replied:
Hello again, @swolchok!

A digression from this PR - can you please give me an example of the following:
_OpenMP does not guarantee that a single per-process intra-op thread pool is going to be used in the application. On the contrary, two different application or inter-op threads may use different OpenMP thread pools for intra-op work._

It'd have been great if the [wiki](https://github.com/pytorch/pytorch/blob/master/docs/source/notes/cpu_threading_torchscript_inference.rst) exemplified it.

Thank you!

swolchok replied:
I don't know much about OpenMP, but I imagine that it has its own thread pools that are separate from the stuff getting touched in this PR.

imaginary-person replied:
Thanks for your response, @swolchok! Yes, it's a digression from this PR.
